Examples/MAX78000/FreeRTOSDemo/FreeRTOS_Debug.c

```diff
@@ -0,0 +1,131 @@
+/********************************************************************************
+ * Copyright (C) 2023 Analog Devices, Inc.
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *  - Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *  - Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *  - Neither the name of Analog Devices, Inc. nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include "FreeRTOS.h"
+#include "task.h"
+#include "tmr.h"
+#include "nvic_table.h"
+
+// Only include the contents of this file if DEBUG == 1
+#if defined(DEBUG) && (DEBUG == 1)
+
+// Scheduler tick is 1ms
+// RTOS Stats Clock should be 10-100x faster
```

Comment:
Have you considered using the RTC instead of a timer instance?  It should have enough resolution, and polling the RTC requires less complete ownership of the peripheral than taking over once of timers.  The roll-over issue is also solved, and it will be more robust against the LP modes

Reply:
[RE: Getting task/context given a PC/memory address inside the HardFault Handler]:
Not that I could find; some linker magic beyond my expertise is required for that bit. With our VS Code environment, the Call Stack does successfully indicate the Task (tested by putting bad memory access inside vTask1):

![VSCode Call Stack](https://github.com/analogdevicesinc/msdk/assets/82959533/ec1952f5-3934-4dee-b437-80941fafe89f)

The Handler itself copies not just the PC but also the CPU register state, and may be useful in the case where a customer wants to use a separate IDE but runs into a similar debug need. But viewing the disassembly is also sufficient given a program counter: 
![VS Code Disassembly View](https://github.com/analogdevicesinc/msdk/assets/82959533/9b01fb85-10ee-46ed-aaea-8d75b99ff1b7)

Reply:
[RE: RTC replacing TMR0] I hadn't considered that, but the rollover issue isn't solved here, it gets worse as far as I can tell. 

The FreeRTOS kernel keeps track of the totalRuntime as a uint32_t variable, without any rollover intelligence. So without customizing the kernel, the GetTimerForStats function should return a uint32_t. The best resolution the RTC gives access to is the SubSeconds, which rolls over every second. So, it gets worse unless I change the kernel around to try and use our custom time_t to track the Stats ticks. Here's it working (note the IDLE task is over 100% -- a common symptom of overflow with the Stats ticks):

![Using RTC as RTOS Stats Timer](https://github.com/analogdevicesinc/msdk/assets/82959533/9d28154b-efbe-4b97-9953-2be5cf622898)

On the low-power point...would it be better to switch to LPTMR0/LPTMR1 instead of TMR0? Or perhaps we can leave a suggestion can be left in the README for this? 

Obviously, this is worse in terms of taking ownership of the application for the Debug build, however it does make the example more robust for implementing LP functionality, and it could point someone in the right direction. A user can still scrap it from application code / use the LPTMR for something else if not debugging since this is in the context of an example and not a part of core SW functionality.

Reply:
> The Handler itself copies not just the PC but also the CPU register state, and may be useful in the case where a customer wants to use a separate IDE but runs into a similar debug need. But viewing the disassembly is also sufficient given a program counter: ![VS Code Disassembly View](https://private-user-images.githubusercontent.com/82959533/317547095-9b01fb85-10ee-46ed-aaea-8d75b99ff1b7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1ODgwMzEsIm5iZiI6MTcxMTU4NzczMSwicGF0aCI6Ii84Mjk1OTUzMy8zMTc1NDcwOTUtOWIwMWZiODUtMTBlZS00NmVkLWFhZWEtOGQ3NWI5OWZmMWI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI4VDAxMDIxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk3ZWJlMDA3MmU2ZGMzZTI1YjhmYzNkNDExYjgzMDZhOGYxMTJjNWQyODQ1MTE3MTM1ODZkODE1ODYwMWM0N2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.G6AcK0wxuFg7300jzvZcaL7pAn_v2M80RtF_VCwha_0)

Ok, no changes needed then.  This would be a good image to put in the README.  If you're feeling real fancy you could manually create a mixed disassembly file for the screenshot.

```shell
arm-none-eabi-objdump -d -S <filename>.elf > disassembly.S
```

> ### [RE: RTC replacing TMR0]
> I hadn't considered that, but the rollover issue isn't solved here, it gets worse as far as I can tell.
> 
> * The FreeRTOS kernel keeps track of the totalRuntime as a uint32_t variable, without any rollover intelligence. So without customizing the kernel, the GetTimerForStats function should return a uint32_t.
> * The best resolution the RTC gives access to is the SubSeconds, which rolls over every second. So, it gets worse unless I change the kernel around to try and use our custom time_t to track the Stats ticks.
> 
> Here's it working (note the IDLE task is over 100% -- a common symptom of overflow with the Stats ticks):
> 
> ![Using RTC as RTOS Stats Timer](https://private-user-images.githubusercontent.com/82959533/317550107-9d28154b-efbe-4b97-9953-2be5cf622898.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1ODgwMzEsIm5iZiI6MTcxMTU4NzczMSwicGF0aCI6Ii84Mjk1OTUzMy8zMTc1NTAxMDctOWQyODE1NGItZWZiZS00Yjk3LTk5NTMtMmJlNWNmNjIyODk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI4VDAxMDIxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg4NTE2ODFlYmQ5OWY2NjEwYmUwMzY3ODQ2YjViMTE5Zjc2MjEwYjNiMzA0YzRjMjg4MWIzZTI2NGUwNTlkN2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hiGymT9m2Ksa9TWx-IRSNbLTfULVw4g4ViRGojsbGuw)
> 
> On the low-power point...would it be better to switch to LPTMR0/LPTMR1 instead of TMR0? Or perhaps we can leave a suggestion in the README for this?
> 
> Obviously, this is worse in terms of taking ownership of the application for the Debug build, however it does make the example more robust for implementing LP functionality, and it could point someone in the right direction. A user can still scrap it from application code / use the LPTMR for something else if not debugging since this is in the context of an example and not a part of core SW functionality.

Ah, I didn't realize it was just a percentage counter and not an actual timestamp.  Might be an interesting kernel mod to use the timespec stuff from `<time.h>` but that's an idea for another day.

I think using an LP timer is better in that it survives LP transitions, but worse because they usually run off of a slower speed clock and don't support 32-bit cascade mode.  So you get less resolution and worse roll-over.

I'll leave it up to you!  :)

Reply:
I'll leave the timer configuration as-is for now. I added some things to the readme to help folks make sense of the extra file. Should be ready! Thanks for the feedback.